internal/pkg/db/redis/data.go

```diff
@@ -458,6 +588,114 @@ func (c *Client) DeleteReadingById(id string) error {
 	return nil
 }
 
+// DeleteReadingsByDevice deletes readings associated with the specified device
+func (c *Client) DeleteReadingsByDevice(device string) error {
+	conn := c.Pool.Get()
+	defer conn.Close()
+
+	ids, err := redis.Strings(conn.Do("ZRANGE", db.ReadingsCollection+":device:"+device, 0, -1))
+	if err != nil {
+		return err
+	}
+
+	err = conn.Send("MULTI")
+	if err != nil {
+		return err
+	}
+
+	for _, id := range ids {
+		err = conn.Send("RENAME", id, db.DeletedReadingsCollection+":"+id)
+		if err != nil {
+			return err
+		}
+	}
+
+	err = conn.Send("EXEC")
+	deleteReadingsChannel <- device
+
+	return nil
+}
+
+// AsyncDeleteReadings Handles the deletion of device readings asynchronously. This function is expected to be running
+// in a go-routine and works with the "DeleteReadingsByDevice" function for better performance.
+func (c *Client) AsyncDeleteReadings(loggingClient logger.LoggingClient) {
+	loggingClient.Debug("Starting background event deletion process")
+	for {
+		select {
+		case device, ok := <-deleteReadingsChannel:
+			if ok {
+				loggingClient.Debug("Deleting reading data for device: " + device)
+				startTime := time.Now()
+				c.deleteRenamedReadings(device, loggingClient)
+				loggingClient.Debug(fmt.Sprintf("Deleted readings for device: '%s', elapsed time: %s", device, time.Since(startTime)))
+			}
```

Comment:
See the comment about closing the channel above. If a read happens when the channel is closed, then ok will be false and this method should explicitly return -- which will, in turn, free up resources assigned to the goroutine.